---
title: "Result Part2: Cell subtype separation"
output: html_notebook
---

* Data downloaded from GEO with `GSE75748`: [click here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75748)
    + 4 files: (cell type, time course)
    + **Cell type**
        - 19 bulk samples, 1018 single cells
        - 7 cell types: H1, H9, EC, NPC, DEC, HFF, TB
    + **Time course data**
        - 15 bulk samples, 758 single cells        

* Analysis
    + PCA plots colored by cell type
    + heatmap plots of expression level of different cell types


## Chu Cell Type Dataset Analysis
### Data preprocessing

```{r}
bulk_ct <- read.csv("Raw_data/Chu/GSE75748_bulk_cell_type_ec.csv", header = T)
bulk_tc <- read.csv("Raw_data/Chu/GSE75748_bulk_time_course_ec.csv", header = T)
sc_ct <- read.csv('Raw_data/Chu/GSE75748_sc_cell_type_ec.csv', header = T)
sc_tc <- read.csv("Raw_data/Chu/GSE75748_sc_time_course_ec.csv", header = T)
# fitlering (remove > 90% unexpressed genes)
gene_idx = apply(sc_ct[,-1], 1, function(x) sum(x>0))>0.1*ncol(sc_ct)
sum(gene_idx)
rownames(bulk_ct) <- bulk_ct$X; bulk_ct <- bulk_ct[,-1]
rownames(sc_ct) <- sc_ct$X;sc_ct <- sc_ct[gene_idx, -1];
gene_idx = apply(sc_tc[,-1], 1, function(x) sum(x>0))>0.1*ncol(sc_tc)
sum(gene_idx)
rownames(bulk_tc) <- bulk_tc$X;bulk_tc <- bulk_tc[,-1]
rownames(sc_tc) <- sc_tc$X;sc_tc <- sc_tc[gene_idx, -1];
save(bulk_ct,bulk_tc, sc_ct, sc_tc,file = 'Data/Chu_fil.RData')

```

### PCA plots
```{r}
library(tidyverse)
library(broom)
library(ggfortify) # for plotting pca
library(ggsci)
data = readRDS('Data/Chu_5datalist.rds')

# add MAGIC and DCA
chu_magic = read.csv('Data/Chu_celltype_magic.csv')
chu_magic = t(chu_magic[,-1]);dimnames(chu_magic) = dimnames(data[[2]])
data[[length(data)+1]] = chu_magic; names(data)[length(data)] = "MAGIC"

chu_dca = read.csv("Data/Chu_celltype_dca.tsv", sep = "\t")
rownames(chu_dca) = chu_dca$X;chu_dca = chu_dca[,-1] # after rounding , PLGLB2 has 0 for all cells, DCA removed this gene
PLGLB2.idx =  which(rownames(data[[2]])=="PLGLB2")
chu_dca = rbind(chu_dca[1:(PLGLB2.idx-1),], rep(0, ncol(chu_dca)), chu_dca[PLGLB2.idx,])
rownames(chu_dca)[PLGLB2.idx] = "PLGLB2"
# chu_dca = chu_dca[,-1]
data[[length(data)+1]] = chu_dca; names(data)[length(data)] = "DCA"

saveRDS(data, file = 'Data/Chu_7datalist.rds')
```

```{r}
data = readRDS("Data/Chu_7datalist.rds")
names(data)
# format the name
names(data) = c('bulk', 'Raw', "G2S3", "scImpute", "VIPER", "MAGIC", "DCA")
#reorder the methods: G2S3, SAVER, MAGIC, scImpute, VIPER, DCA
data = data[c(1:3, 6, 4, 5, 7)]


plot_marker = function(sc_ct, method){
	df = (sc_ct[c("GATA6", "NANOG"),] + 0.1) %>% t() %>% log()%>%data.frame()
	df$celltype = as.factor(substring(colnames(sc_ct),first = 1, last = 2))
	p = ggplot(df, aes(x=GATA6, y=NANOG, color=celltype)) + geom_point()+ scale_color_npg() + theme_bw(base_size = 16) + ggtitle(method) +theme(plot.title = element_text(hjust = 0.5, size = 20, face = "bold")) + labs(color = "Cell Type")
	return(p)
	}
plot_list = lapply(2:length(data), function(i) {plot_marker(data[[i]], names(data)[i])})

library(gridExtra)
pdf('Plots/Figure_S3.pdf',width=16, height=8)
do.call("grid.arrange", c(plot_list, ncol=3))
dev.off()

# plot first 1, 3 pcs
plot_pca = function(sc_ct, method){
	print(method)
	cg = rowSums(sc_ct)
	sc_ct[sc_ct<0]=0
	cg.idx = rowSums(sc_ct)>0 # remove genes with all zero reads
	sc_pca <-(sc_ct[cg.idx,] + 0.1) %>% t() %>% log()%>% scale() %>% prcomp()  
	sc_pca = data.frame(sc_pca$x[,1:5])
	sc_pca$celltype = as.factor(substring(colnames(sc_ct),first = 1, last = 2))

	p = ggplot(sc_pca, aes(x=PC1, y=PC3, color=celltype)) + geom_point()+ scale_color_npg() + theme_bw(base_size = 16) + ggtitle(method) +theme(plot.title = element_text(hjust = 0.5, size = 20, face = "bold")) + labs(color = "cell type")
	return(p)
}

plot_list = lapply(2:length(data), function(i) {plot_pca(data[[i]], names(data)[i])})
names(plot_list) <- names(data)[-1]
saveRDS(plot_list, file = "Data/pca_plot.rds")


```

## Plot of Figure 2

```{r}
plot_list = readRDS('Data/pca_plot.rds')
library(gridExtra)
pdf('Plots/Figure_2.pdf',width=16, height=8)
do.call("grid.arrange", c(plot_list, ncol=3))
dev.off()


```



## Plot of Figure 2 (new 03/17/2020)

```{r}
data = readRDS("Data/Chu_7datalist.rds")
names(data)
# format the name
names(data) = c('bulk', 'raw', "G2S3", "scImpute", "VIPER", "MAGIC", "DCA")
#reorder the methods: G2S3, SAVER, MAGIC, scImpute, VIPER, DCA
data = data[c(1:3, 6, 4, 5, 7)]


library(fpc)
library(tidyverse)
calc_10pc = function(sc_ct){
	cg = rowSums(sc_ct)
	sc_ct[sc_ct<0]=0
	cg.idx = rowSums(sc_ct)>0 # remove genes with all zero reads
	sc_pca <-(sc_ct[cg.idx,] + 0.1) %>% t() %>% log()%>% scale() %>% prcomp()  
	sc_pca = data.frame(sc_pca$x[,1:10])
	sc_pca$celltype = as.factor(substring(colnames(sc_ct),first = 1, last = 2))
	return(sc_pca)
}

pcs_list = lapply(2:length(data), function(i) {calc_10pc(data[[i]])})
names(pcs_list) = names(data)[-1]
saveRDS(pcs_list, file = "Data/methods_10pcs.rds")

# The plot
calc_DunnIndex <- function(pc_df){
	dunn =c()
	intr_inter = c()
	avg.sil = c()
	for(i in 1:10){
		d = dist(pc_df[,1:i])
		c = as.integer(pc_df[,11])
		c.stat = cluster.stats(d, c)
		dunn = c(dunn, c.stat$dunn)
		intr_inter = c(intr_inter, c.stat$average.between/c.stat$average.within)
		avg.sil = c(avg.sil, c.stat$avg.silwidth)
	}
	return(list(dunn = dunn, intr_inter = intr_inter, avg.sil = avg.sil))
}

cluster.stats = lapply(pcs_list, function(x) calc_DunnIndex(x))
saveRDS(cluster.stats, file = "Data/cluster_stats.rds")




cluster.stats = readRDS("Data/cluster_stats.rds")
names(cluster.stats)[1] = "Raw"
#plot figures
library(ggsci)
library(ggplot2)

gen_cluster_stat_plot = function(measure){

	if(measure == "intr_inter"){yl = "Average Inter/Intra-subtype Distance Ratio"}else{
		yl = "Silhouette Coefficient"}

	data_list= lapply(cluster.stats, function(x) get(measure, x))
	df = do.call(cbind, data_list)
	library(reshape2)
	df.melt <- melt(df); colnames(df.melt)[1:2]= c("PCs", "Method")
	library(ggsci)
	library(ggplot2)
	p1 = ggplot(df.melt, aes(x = PCs, y = value, color = Method)) +theme_bw() +
	geom_line() + ylab(yl) + scale_color_lancet() + xlab("Number of PCs") + scale_x_continuous(breaks=seq(1:10))
	return(p1)
	# plotname = paste0("./Plots/Cluster_stat_",measure, ".pdf")
	# ggsave(p1, file = plotname)
}

f1 = gen_cluster_stat_plot("intr_inter")
f2 = gen_cluster_stat_plot("avg.sil")
library(gridExtra)
fig = grid.arrange(f1, f2)
ggsave(fig, file = "./Plots/FIgure2.pdf", width = 10)
```




## plot of PC1 vs PC2 (Figure S2); 

```{r}
# plot first 1, 2 pcs
pcs_list = readRDS("./Data/methods_10pcs.rds")
names(pcs_list)[1] = "Raw"
plot_pca = function(sc_pca, method){
	print(method)
	p = ggplot(sc_pca, aes(x=PC1, y=PC2, color=celltype)) + geom_point()+ scale_color_npg() + theme_bw(base_size = 16) + ggtitle(method) +theme(plot.title = element_text(hjust = 0.5, size = 20, face = "bold"))+ labs(color = "Cell Type")
	return(p)
}

plot_list = lapply(1:length(pcs_list), function(i) {plot_pca(pcs_list[[i]], names(pcs_list)[i])})

library(gridExtra)
library(ggsci)
pdf('Plots/Figure_S2.pdf',width=16, height=8)
do.call("grid.arrange", c(plot_list, ncol=3))
dev.off()

plot_pca = function(sc_pca, method){
	print(method)
	p = ggplot(sc_pca, aes(x=PC1, y=PC3, color=celltype)) + geom_point()+ scale_color_npg() + theme_bw(base_size = 16) + ggtitle(method) +theme(plot.title = element_text(hjust = 0.5, size = 20, face = "bold"))+ labs(color = "Cell Type")
	return(p)
}

plot_list = lapply(1:length(pcs_list), function(i) {plot_pca(pcs_list[[i]], names(pcs_list)[i])})
pdf('Plots/Figure3.pdf',width=16, height=8)
do.call("grid.arrange", c(plot_list, ncol=3))
dev.off()




```








